blueprints/varnish-support.md

```diff
@@ -0,0 +1,117 @@
+<!--
+Licensed to the Apache Software Foundation (ASF) under one
+or more contributor license agreements.  See the NOTICE file
+distributed with this work for additional information
+regarding copyright ownership.  The ASF licenses this file
+to you under the Apache License, Version 2.0 (the
+"License"); you may not use this file except in compliance
+with the License.  You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing,
+software distributed under the License is distributed on an
+"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+KIND, either express or implied.  See the License for the
+specific language governing permissions and limitations
+under the License.
+-->
+# Add Varnish Cache Support
+
+## Problem Description
+
+Currently Traffic Control uses Traffic Server as the underlying cache server. We can expand that by introducing Varnish cache as an option for the cache server used with its great performance, robustness and modularity.
+
+## Proposed Change
+
+From a high level point of view, ATS operates based on configuration files that describe in details how it should work and interact with other servers in the cache hierarchy. These configuration files are managed and generated using `t3c` components that utilize Traffic Ops APIs to get profiles and parameters data required for the configuration files. The proposed change is to use the same data fetched from Traffic Ops APIs to generate configuration files for Varnish cache with almost the same functionality.
+
+Note: the changes should not affect existing components but rather build on them.
+
+### Traffic Portal Impact
+
+n/a
+
+### Traffic Ops Impact
+
+n/a
+
+#### REST API Impact
+
+n/a
+
+#### Client Impact
+
+n/a
+
+#### Data Model / Database Impact
+
+- A new profile type will be added `VARNISH_PROFILE` but it won't affect the model or other structures in `lib/go-tc`
+- `DeliveryService` structs contain fields related to ATS like `remapText`. It will be parsed and translated to Varnish configuration.
+
+### ORT Impact
+
+- `go-varnish` package will be developed to handle generating configuration files for Varnish, Hitch, `varnishncsa` and `logrotate`. For detailed description of mapping configuration files from ATS to Varnish refer to [Varnish Support](https://github.com/apache/trafficcontrol/wiki/Varnish-Support) wiki.
```

Comment:
I would love it, personally, if we stopped doing the whole `go-` prefix thing. It's a Go package, so it kinda seems redundant.

Also, I'd vote we make it a subpackage of the atscfg package. Or maybe that would be a big refactor to pull generic stuff out into like a `genconfig` package with separate sub-packages for different caching systems. Might be out-of-scope, but you do mention a refactor below.

Reply:
While it is definitely a better and cleaner solution to have some common package between all cache generation packages, it might be a out of scope as the refactoring mentioned was meant to be about making some functions related to config data public and reusable from another package but not a new package containing every thing related to config data. But, It is an interesting idea and we could discuss if it should be a priority or another project to be done later.

Not sure exactly what is meant by making `varnishcfg` subpackage to `atscfg`? Are new packages for other cache servers will be also part of `atscfg`?

Reply:
no, each cache server software should be its own package, but just ideally all underneath a larger generic package that defers to specific implementations to calculate file contents. Which may be out of scope, like I said.